<a id="toc"></a>
<h1 style="background-color:lightblue;font-size:275%;text-align:center;border-radius:50px;">Table of Contents</h1>

* [1. Introduction](#1)
    * [1.1 Libraries](#1.1)
    * [1.2 Data Loading](#1.2)
    * [1.3 Data Cleaning](#1.3)
* [2. Exploratory Data Analysis (EDA)](#2)
    * [2.1 Unique Value Counts & Observations](#2.1)
    * [2.2 Repeat Addresses](#2.2)
    * [2.3 States Without Dunkin's](#2.3)
    * [2.4 Mapping Store Locations](#2.4)
    * [2.5 Top 10 States](#2.5)
    * [2.6 Top 10 Cities](#2.6)

<a id="1"></a>
<h2 style="font-size:250%;text-align:left;border-radius:50px;">Introduction</h2>

<a id="1.1"></a>
<h3 style="font-size:200%;text-align:left;border-radius:50px;">Libraries</h3>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

<a id="1.2"></a>
<h3 style="font-size:200%;text-align:left;border-radius:50px;">Data Loading</h3>

In [2]:
df = pd.read_csv('../input/dunkin-locations/dunkin_stores.csv')
df.head(10)

row_id          address_line_1 address_line_2         city state    zip  \
0       0           1114 1st St N      Suite 200    Alabaster    AL  35007   
1       1     8171 US Highway 431            NaN  Albertville    AL  35950   
2       2    1700 US Highway 72 E            NaN       Athens    AL  35611   
3       3        171 N College St            NaN       Auburn    AL  36830   
4       4       2049 S College St            NaN       Auburn    AL  36832   
5       5  2178 Eastern Valley Rd            NaN     Bessemer    AL  35022   
6       6        1820 Gadsden Hwy        Ste 100   Birmingham    AL  35235   
7       7          2109 6th Ave S            NaN   Birmingham    AL  35233   
8       8          300 Commons Dr            NaN   Birmingham    AL  35209   
9       9     400 Cahaba Park Cir        Suite 1   Birmingham    AL  35242   

            phone            mon_hrs            tue_hrs            wed_hrs  \
0  (205) 624-4880  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM   
1  (256) 869-6075  5:00 AM - 9:00 PM  5:00 AM - 9:00 PM  5:00 AM - 9:00 PM   
2  (256) 960-9302  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM   
3  (334) 707-8552  6:00 AM - 3:00 PM  6:00 AM - 3:00 PM  6:00 AM - 3:00 PM   
4  (334) 501-2233  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM   
5  (205) 425-1333  6:00 AM - 4:00 PM  6:00 AM - 4:00 PM  6:00 AM - 4:00 PM   
6  (205) 533-8295  6:00 AM - 7:00 PM  6:00 AM - 7:00 PM  6:00 AM - 7:00 PM   
7  (205) 297-8777  6:00 AM - 6:00 PM  6:00 AM - 6:00 PM  6:00 AM - 6:00 PM   
8  (205) 518-5250  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM  5:00 AM - 8:00 PM   
9  (205) 502-7311  5:00 AM - 6:00 PM  5:00 AM - 6:00 PM  5:00 AM - 6:00 PM   

   ...             sat_hrs            sun_hrs drive-thru mobile-order  \
0  ...   6:00 AM - 8:00 PM  6:00 AM - 8:00 PM       True        False   
1  ...  5:00 AM - 10:00 PM  5:00 AM - 9:00 PM       True         True   
2  ...   5:00 AM - 8:00 PM  5:00 AM - 8:00 PM       True         True   
3  ...   6:00 AM - 3:00 PM  6:00 AM - 3:00 PM      False        False   
4  ...   5:00 AM - 8:00 PM  6:00 AM - 8:00 PM       True         True   
5  ...   6:00 AM - 4:00 PM  7:00 AM - 4:00 PM       True         True   
6  ...   6:00 AM - 7:00 PM  6:00 AM - 7:00 PM       True         True   
7  ...   6:00 AM - 4:00 PM  7:00 AM - 5:00 PM       True         True   
8  ...   6:00 AM - 8:00 PM  6:00 AM - 8:00 PM       True         True   
9  ...   6:00 AM - 6:00 PM  6:00 AM - 6:00 PM       True        False   

   dunkin-card   kcup  curbside-pickup  has-baskin-robbins    loc_lat  \
0         True   True            False               False        NaN   
1         True  False            False                True        NaN   
2         True   True            False               False  34.781698   
3         True   True            False               False  32.608800   
4         True   True            False               False  32.567323   
5         True   True            False               False  33.334624   
6         True   True            False               False        NaN   
7         True   True            False               False  33.507720   
8         True   True            False               False  33.441958   
9         True   True            False               False        NaN   

    loc_long  
0        NaN  
1        NaN  
2 -86.934524  
3 -85.481849  
4 -85.501589  
5 -86.995903  
6        NaN  
7 -86.798970  
8 -86.830757  
9        NaN  

[10 rows x 22 columns]

<a id="1.3"></a>
<h3 style="font-size:200%;text-align:left;border-radius:50px;">Data Cleaning</h3>

In [3]:
print(df.info())
print('-------------------------')
print(df.shape)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9248 entries, 0 to 9247
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   row_id              9248 non-null   int64  
 1   address_line_1      9248 non-null   object 
 2   address_line_2      1763 non-null   object 
 3   city                9248 non-null   object 
 4   state               9248 non-null   object 
 5   zip                 9248 non-null   int64  
 6   phone               9126 non-null   object 
 7   mon_hrs             9248 non-null   object 
 8   tue_hrs             9248 non-null   object 
 9   wed_hrs             9248 non-null   object 
 10  thu_hrs             9248 non-null   object 
 11  fri_hrs             9248 non-null   object 
 12  sat_hrs             9248 non-null   object 
 13  sun_hrs             9248 non-null   object 
 14  drive-thru          9248 non-null   bool   
 15  mobile-order        9248 non-null   bool   
 16  dunkin

In [4]:
df[["address_line_1", "address_line_2", "city", "state", "zip", "phone"]] = \
df[["address_line_1", "address_line_2", "city", "state", "zip", "phone"]].astype('string')

print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9248 entries, 0 to 9247
Data columns (total 22 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   row_id              9248 non-null   int64  
 1   address_line_1      9248 non-null   string 
 2   address_line_2      1763 non-null   string 
 3   city                9248 non-null   string 
 4   state               9248 non-null   string 
 5   zip                 9248 non-null   string 
 6   phone               9126 non-null   string 
 7   mon_hrs             9248 non-null   object 
 8   tue_hrs             9248 non-null   object 
 9   wed_hrs             9248 non-null   object 
 10  thu_hrs             9248 non-null   object 
 11  fri_hrs             9248 non-null   object 
 12  sat_hrs             9248 non-null   object 
 13  sun_hrs             9248 non-null   object 
 14  drive-thru          9248 non-null   bool   
 15  mobile-order        9248 non-null   bool   
 16  dunkin

<a id="2"></a>
<h2 style="font-size:250%;text-align:left;border-radius:50px;">Exploratory Data Analysis (EDA)</h2>

<a id="2.1"></a>
<h3 style="font-size:200%;text-align:left;border-radius:50px;">Unique Value Counts & Observations</h3>

In [5]:
df.nunique()

row_id                9248
address_line_1        9144
address_line_2         970
city                  3126
state                   45
zip                   5062
phone                 8989
mon_hrs                235
tue_hrs                230
wed_hrs                227
thu_hrs                230
fri_hrs                231
sat_hrs                229
sun_hrs                229
drive-thru               2
mobile-order             2
dunkin-card              2
kcup                     2
curbside-pickup          2
has-baskin-robbins       2
loc_lat               6095
loc_long              6095
dtype: int64

Some observations from this readout:
* From this readout we can see that while there are 9248 stores located in the dataset, there are only 9144 unique addresses listed. This suggests that some addresses have more than one Dunkin' on site. These are likely large public areas such as airports or stadiums. To further investigate, we can look at the addresses that appear more than once in the set.
* There are 5 states that do not contain a single Dunkin' store. At least, not one listed on the Dunkin' website. Finding out which states those are should be very easy. Seeing as Dunkin' [started in Massachusetts](https://www.thebalancesmb.com/the-history-of-dunkin-donuts-3973232), I would expect all east coast states to have locations, which suggests that either some midwest or west coast states might not have Dunkin's. I would expect Washington to be one of the states, as that is where Starbucks, Dunkin's main competitor, was founded.
* Many addresses weren't able to be found with the geolocator, leaving roughly 1/3 of the addresses without latitude and longitude data.

<a id="2.2"></a>
<h4 style="font-size:200%;text-align:left;border-radius:50px;">Repeat Addresses</h4>

In [6]:
df.address_line_1.value_counts().head(20)

JFK Intl Airport                  14
Logan Airport                      8
Newark Int'L Airport               6
Ronald Reagan Airport              6
Dfw Int. Airport                   5
Philadelphia Int. Airport          4
4 Yawkey Way                       3
Bwi Airport                        3
Dulles Intl Airport                3
75 Main St                         3
Hartsfield Jackson Atl Airport     3
Dallas Love Field Airport          3
2 Main St                          3
339 Main St                        3
Wichita Airport                    3
1 Lincoln Financial Field Way      3
2 Penn Plz                         3
1189 Green St                      2
196 Main St                        2
217 Main St                        2
Name: address_line_1, dtype: Int64

As expected the list consists of airports and other large venues like sports stadiums. For example, *1 Lincoln Financial Field Way* is the address of Lincoln Financial Field, where the Philadelphia Eagles play.

<a id="2.3"></a>
<h4 style="font-size:200%;text-align:left;border-radius:50px;">States without Dunkin's</h4>

In [7]:
states = {
        'AK': 'Alaska',
        'AL': 'Alabama',
        'AR': 'Arkansas',
        'AZ': 'Arizona',
        'CA': 'California',
        'CO': 'Colorado',
        'CT': 'Connecticut',
        'DC': 'District of Columbia',
        'DE': 'Delaware',
        'FL': 'Florida',
        'GA': 'Georgia',
        'HI': 'Hawaii',
        'IA': 'Iowa',
        'ID': 'Idaho',
        'IL': 'Illinois',
        'IN': 'Indiana',
        'KS': 'Kansas',
        'KY': 'Kentucky',
        'LA': 'Louisiana',
        'MA': 'Massachusetts',
        'MD': 'Maryland',
        'ME': 'Maine',
        'MI': 'Michigan',
        'MN': 'Minnesota',
        'MO': 'Missouri',
        'MS': 'Mississippi',
        'MT': 'Montana',
        'NC': 'North Carolina',
        'ND': 'North Dakota',
        'NE': 'Nebraska',
        'NH': 'New Hampshire',
        'NJ': 'New Jersey',
        'NM': 'New Mexico',
        'NV': 'Nevada',
        'NY': 'New York',
        'OH': 'Ohio',
        'OK': 'Oklahoma',
        'OR': 'Oregon',
        'PA': 'Pennsylvania',
        'RI': 'Rhode Island',
        'SC': 'South Carolina',
        'SD': 'South Dakota',
        'TN': 'Tennessee',
        'TX': 'Texas',
        'UT': 'Utah',
        'VA': 'Virginia',
        'VT': 'Vermont',
        'WA': 'Washington',
        'WI': 'Wisconsin',
        'WV': 'West Virginia',
        'WY': 'Wyoming'
}

In [8]:
for state_abrv, state_full in states.items():
    if state_abrv not in df.state.unique():
        print(state_full)

Idaho
Montana
North Dakota
Oregon
South Dakota
Washington


This output is a little unexpected. As noted above, there are 45 distinct values for the `state` column in the dataframe, suggesting that only 5 states should be included in the above readout. Instead, we got 6. Where is the disconnect? This likely stems from the fact that while the District of Columbia is a valid entry in the `states` dictionary, it is not included in the base 50 states. So, there are actually 6 states without a Dunkin' store! We can also note that our assumption about Washington state was correct, and that the other states do tend towards the upper midwest and west coast as predicted.

We can confirm this suspicion by checking the values that the `state` variable takes on within the dataset. If 'DC' is included, we know this hunch to be true.

In [9]:
df.state.unique().__contains__('DC')

True

<a id="2.4"></a>
<h4 style="font-size:200%;text-align:left;border-radius:50px;">Mapping Store Locations</h4>

Let's take a quick look at the general density of stores in each state.

In [10]:
# function to change state abbreviation to state name using predefined state dict
def get_val(key_to_find):
    for key, val in states.items():
        if key_to_find == key:
            return val
    return 'Key does not exist'

store_state = df.groupby('state').count()
store_state['state_name'] = [get_val(state_abrv) for state_abrv in store_state.index]

fig = px.choropleth(store_state,
                    locations = store_state.index,
                    color = 'row_id',
                    color_continuous_scale = 'oranges',
                    hover_name = store_state.state_name,
                    locationmode = 'USA-states',
                    labels = {'row_id': 'Number of Stores'}
                    )

fig.update_layout(title_text = "Count of Dunkin' stores per state",
                  geo_scope = 'usa')

fig.show()

In [11]:
fig = go.Figure(data = go.Scattergeo(
        lon = df['loc_long'],
        lat = df['loc_lat'],
        text = df['address_line_1'] + ', ' + df['city'] + ', ' + df['state'] + ', ' + df['zip'],
        mode = 'markers',
        marker_color = 'orange'))

fig.update_layout(title = 'Dunkin Store Locations',
                  geo_scope = 'usa')

fig.show()

As shown by the above maps, the east coast has a much higher density of Dunkin' stores compared to the rest of the country. There is also one erronious store mapped in the upper left-hand corner of the second map. As we know, there are no Dunkin' locations in Washington, so this must be an error in the geocoding of the latitude and longitude data. Scrolling over the point confirms this, as the address listed in is Massachusetts.

<a id="2.5"></a>
<h4 style="font-size:200%;text-align:left;border-radius:50px;">Top 10 States</h4>

In [12]:
df_top_states = df.groupby('state').count().sort_values(by = 'row_id', ascending = False).head(10)

fig = px.bar(df_top_states, 
             x = df_top_states.index, 
             y = 'row_id',
             labels = {'row_id': 'Number of Locations', 'state': 'State'},
             template = 'plotly_white')

fig.update_layout(title = 'Top 10 States by Number of Dunkin Locations')

fig.update_traces(marker_color = 'orange')

fig.show()

<a id="2.6"></a>
<h4 style="font-size:200%;text-align:left;border-radius:50px;">Top 10 Cities</h4>

In [13]:
df_top_cities = df.groupby('city').count().sort_values(by = 'row_id', ascending = False).head(10)

fig = px.bar(df_top_cities, 
             x = df_top_cities.index, 
             y = 'row_id',
             labels = {'row_id': 'Number of Locations', 'city': 'City'},
             template = 'plotly_white')

fig.update_layout(title = 'Top 10 Cities by Number of Dunkin Locations')

fig.update_traces(marker_color = 'orange')

fig.show()